File: Employee_UI/employee-management-ui/src/app/core/services/holiday.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Holiday, CreateHolidayRequest, UpdateHolidayRequest, HolidayResponse } from '../models/holiday.model';

@Injectable({ providedIn: 'root' })
export class HolidayService {

  private BASE_URL = 'http://localhost:8080/api/holidays';

  constructor(private http: HttpClient) {}

  // ============================
  // GET HOLIDAYS BY YEAR
  // ============================
  getHolidaysByYear(year: number): Observable<Holiday[]> {
    return this.http.get<Holiday[]>(`${this.BASE_URL}/year/${year}`);
  }

  // ============================
  // CREATE HOLIDAY (ADMIN ONLY)
  // ============================
  createHoliday(data: CreateHolidayRequest): Observable<HolidayResponse> {
    return this.http.post<HolidayResponse>(this.BASE_URL, data);
  }

  // ============================
  // UPDATE HOLIDAY (ADMIN ONLY)
  // ============================
  updateHoliday(id: number, data: UpdateHolidayRequest): Observable<HolidayResponse> {
    return this.http.put<HolidayResponse>(`${this.BASE_URL}/${id}`, data);
  }

  // ============================
  // DELETE HOLIDAY (ADMIN ONLY)
  // ============================
  deleteHoliday(id: number): Observable<any> {
    return this.http.delete(`${this.BASE_URL}/${id}`);
  }

   getHolidaysByMonth(year: number, month: number): Observable<Holiday[]> {
    return this.http.get<Holiday[]>(
      `${this.BASE_URL}/year/${year}/month/${month}`
    );
  }
}

File: JPA_TASK_01/src/main/java/com/example/JPA_TASK_01/Service/HolidayService.java
package com.example.JPA_TASK_01.Service;

import com.example.JPA_TASK_01.DTO.Holiday.HolidayRequest;
import com.example.JPA_TASK_01.DTO.Holiday.HolidayResponse;

import java.util.List;

public interface HolidayService {

    HolidayResponse createHoliday(HolidayRequest request);

    HolidayResponse updateHoliday(Integer holidayId, HolidayRequest request);

    void deleteHoliday(Integer holidayId);

    List<HolidayResponse> getHolidaysByYear(int year);

    List<HolidayResponse> getHolidaysByYearAndMonth(int year, int month);
}
